core/vote/vote_manager.go

```diff
@@ -0,0 +1,217 @@
+			if ev == nil {
+				continue
+			}
+			switch ev.Data.(type) {
+			case downloader.StartEvent:
+				startVote = false
+			case downloader.FailedEvent:
+				startVote = false
+			case downloader.DoneEvent:
+				startVote = true
+			}
+		case cHead := <-voteManager.chainHeadCh:
+			if !startVote || cHead.Block == nil {
+				continue
+			}
+
+			curHead := cHead.Block.Header()
+			// Check if cur validator is within the validatorSet at curHead
+			if !voteManager.engine.WithinValidatorSet(voteManager.chain, curHead) {
+				continue
+			}
+
+			// Vote for curBlockHeader block.
+			vote := &types.VoteData{
+				TargetNumber: curHead.Number.Uint64(),
+				TargetHash:   curHead.Hash(),
+			}
+			voteMessage := &types.VoteEnvelope{
+				Data: vote,
+			}
+
+			// Put Vote into journal and VotesPool if we are active validator and allow to sign it.
+			if ok, sourceNumber, sourceHash := voteManager.UnderRules(curHead); ok {
+				if sourceHash == (common.Hash{}) {
+					continue
+				}
+
+				voteMessage.Data.SourceNumber = sourceNumber
+				voteMessage.Data.SourceHash = sourceHash
+
+				if err := voteManager.signer.SignVote(voteMessage); err != nil {
+					log.Debug("Failed to sign vote", "err", err)
+					votesSigningErrorMetric(vote.TargetNumber, vote.TargetHash).Inc(1)
+					continue
+				}
+				if err := voteManager.journal.WriteVote(voteMessage); err != nil {
+					log.Warn("Failed to write vote into journal", "err", err)
+					votesJournalErrorMetric(vote.TargetNumber, vote.TargetHash).Inc(1)
+					continue
+				}
+
+				log.Info("vote manager produced vote", "voteHash", voteMessage.Hash())
+				voteManager.pool.PutVote(voteMessage)
+				votesManagerMetric(vote.TargetNumber, vote.TargetHash).Inc(1)
+			}
+		case <-voteManager.chainHeadSub.Err():
+			return
+		}
+	}
+}
+
+// UnderRules checks if the produced header under the following rules:
+// A validator must not publish two distinct votes for the same height. (Rule 1)
+// A validator must not vote within the span of its other votes . (Rule 2)
+// Validators always vote for their canonical chain’s latest block. (Rule 3)
+func (voteManager *VoteManager) UnderRules(header *types.Header) (bool, uint64, common.Hash) {
+	justifiedHeader := voteManager.getHighestJustifiedHeader(voteManager.chain, header)
+	if justifiedHeader == nil {
+		log.Error("highestJustifiedHeader is nil")
+		return false, 0, common.Hash{}
+	}
+
+	sourceNumber := justifiedHeader.Number.Uint64()
+	sourceHash := justifiedHeader.Hash()
+	targetNumber := header.Number.Uint64()
+
+	journal := voteManager.journal
+	walLog := journal.walLog
+
+	firstIndex, err := walLog.FirstIndex()
+	if err != nil {
+		log.Error("Failed to get firstIndex of vote journal", "err", err)
+		return false, 0, common.Hash{}
+	}
+
+	lastIndex, err := walLog.LastIndex()
+	if err != nil {
+		log.Error("Failed to get lastIndex of vote journal", "err", err)
+		return false, 0, common.Hash{}
+	}
+
+	for index := lastIndex; index >= firstIndex; index-- {
```

Comment:
In most cases, we have to iterate from `lastIndex` to `firstIndex`, can we stop it earlier? Like if `vote.Data.TargetNumber` is smaller than the current `targetNumber`, we can return true? Because we never vote to a Target that is less or equal to ever voted target?

Reply:
Won' the case that vote.Data.TargetNumber < targetNumber but vote.Data.SourceNumber >sourceNumber happen? that is still outside our rules, right?

Reply:
Updated